janus_server/src/datastore.rs

```diff
@@ -109,8 +111,26 @@ impl<C: Clock> Datastore<C> {
         for<'a> F:
             Fn(&'a Transaction<C>) -> Pin<Box<dyn Future<Output = Result<T, Error>> + Send + 'a>>,
     {
-        // Open transaction.
-        let mut client = self.pool.get().await?;
+        let backoff = ExponentialBackoff {
+            initial_interval: StdDuration::from_secs(1),
+            max_interval: StdDuration::from_secs(30),
+            multiplier: 2.0,
+            max_elapsed_time: Some(StdDuration::from_secs(600)),
+            ..Default::default()
+        };
+
+        // Open transaction, retrying if we encounter timeouts or connection refused errors (which
+        // occur if the Cloud SQL Proxy hasn't started yet and manifest as `PoolError::Backend`).
+        let mut client = retry(backoff, || async {
+            self.pool.get().await.map_err(|error| match error {
+                PoolError::Timeout(_) | PoolError::Backend(_) => {
```

Comment:
Unfortunately, the error we get for `ECONNREFUSED` doesn't materialize as a `DbError`. I believe it'll be an [`ErrorInner`](https://docs.rs/tokio-postgres/0.7.6/src/tokio_postgres/error/mod.rs.html#366) with `kind: Kind::Connect`, but neither the `kind` field nor the `Kind` enum are public, so we can't match on the error type. I think I'll do as you suggest, and do a startup check instead of putting retries on every single call to `self.pool.get()`.

Reply:
This is tricky. The error we get if the connection is refused because the sidecar isn't up yet looks like:
`Backend(Error { kind: Connect, cause: Some(Os { code: 111, kind: ConnectionRefused, message: \"Connection refused\" }) })`

The error we get if we provide an unroutable DB address (as the test `aggregator::health_check` does) is:
`Backend(Error { kind: Connect, cause: Some(Custom { kind: Uncategorized, error: "failed to lookup address information: Name or service not known" }) })`

We can't distinguish them programmatically, because the `kind` and `cause` fields are not public. The best I could come up with was to have test configurations give up retrying after 1 second, but to retry for up to a minute otherwise.